salt/file/xmpp/ejabberd_authentication.py

```diff
@@ -0,0 +1,170 @@
+import hmac
+import pathlib
+import struct
+import sys
+from typing import Collection, Mapping, Tuple
+
+# Map from (user, server) to that user's crypted passwords.
+_Config = Mapping[Tuple[bytes, bytes], Collection[str]]
+
+
+def _args():
+    parser = argparse.ArgumentParser(
+        description='External authentication helper for ejabberd.')
+    parser.add_argument(
+        '--config',
+        type=pathlib.Path,
+        required=True,
+        help='Authentication config file.',
+    )
+    parser.add_argument(
+        '--max-passwords-per-user',
+        default=25,
+        type=int,
+        help='Max number of passwords per user.',
+    )
+    return parser.parse_args()
+
+
+def _config(
+    config_path: pathlib.Path,
+    *,
+    max_passwords_per_user: int,
+) -> _Config:
+    raw_config = collections.defaultdict(list)
+    with config_path.open(mode='rb') as config_file:
+        for line in config_file:
+            user, server, crypted_password = line.rstrip(b'\n').split(b':')
+            raw_config[(user, server)].append(crypted_password.decode('utf-8'))
+    config = {}
+    for key, crypted_passwords in raw_config.items():
+        if len(crypted_passwords) > max_passwords_per_user:
+            raise ValueError(f'{key!r} has too many passwords')
+        config[key] = tuple(crypted_passwords[i % len(crypted_passwords)]
+                            for i in range(max_passwords_per_user))
+    return config
+
+
+class _Response(enum.IntEnum):
+    FAILURE = 0
+    SUCCESS = 1
+
+
+def _read() -> Tuple[bytes, bytes]:
+    length_bytes = sys.stdin.buffer.read(2)
+    if not length_bytes:
+        raise EOFError()
+    elif len(length_bytes) != 2:
+        raise ValueError(f'Expected 2 bytes, got {len(length_bytes)}')
+    (length,) = struct.unpack('>H', length_bytes)
+    value_bytes = sys.stdin.buffer.read(length)
+    if len(value_bytes) != length:
+        raise ValueError(f'Expected {length} bytes, got {len(value_bytes)}')
+    operation, _, args = value_bytes.partition(b':')
+    return operation, args
+
+
+def _write(response: _Response) -> None:
+    sys.stdout.buffer.write(struct.pack('>HH', 2, response))
+    sys.stdout.buffer.flush()
+
+
+def _auth(operation_args: bytes, *, config: _Config) -> _Response:
+    user, server, password = operation_args.split(b':', maxsplit=2)
+    try:
+        password_str = password.decode('utf-8')
+    except UnicodeDecodeError:
+        return _Response.FAILURE
+    crypted_passwords = config.get((user, server))
+    if crypted_passwords is None:
+        return _Response.FAILURE
+    password_checks_failure = bytes((int(False),) * len(crypted_passwords))
+    password_checks_actual = bytearray(password_checks_failure)
+    for password_index, crypted_password in enumerate(crypted_passwords):
+        password_checks_actual[password_index] = int(
+            hmac.compare_digest(
+                crypt.crypt(password_str, salt=crypted_password),
+                crypted_password))
```

Comment:
I think this is more readable:

```suggestion
    password_checks_actual = bytearray(
        int(hmac.compare_digest(crypt.crypt(password_str, salt=c), c))
        for c in crypted_passwords);
```

Reply:
Good idea. I don't like `c` as a variable name (even though the scope is really small), so I got rid of the unnecessary `int()` calls to make it all fit.